Add a proper builder for key bindings

Sollace opened the pull request:
This does ~~two~~ three things:
 - it adds proper builder-esque methods for the FabricKeyBinding.Builder class.
 - it adds a list of vanilla keybindings so modders can easily pick and choose which ones are available.
 - expose the identifier mods use to create/register their keybindings so it can be used later.

I'd like to allow custom categories, but I don't know where the warning on "unknown key categories" comes from so I can't change that.

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +27,87 @@
-			this.binding = binding;
+		private InputUtil.Type type = InputUtil.Type.KEYSYM;
+
+		private Identifier id = new Identifier("fabric", "unnamed");
```
This should be initialized as `null` and throw an exception if built with no defined id.

Sollace replied:
Already working on that. :3

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +27,87 @@
+		private String category = KeyCategory.MISC;
+
+		private Builder() {
+
```
Redundant whitespace

Sollace replied:
Maybe redundant, may also be stylistic. Do you want it on one line?

mattmess1221 commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +27,87 @@
 
 		public FabricKeyBinding build() {
-			return binding;
+			return new FabricKeyBinding(id, type, code, category);
```
You should be checking if everything has been set first. Throw `IllegalStateExeption` if it hasn't.

Sollace replied:
That might be a good idea. I'll put in checks for the required stuff.

mattmess1221 replied:
If you're going to use `NullPointerException`,  you'd might as well use `Objects.requireNonNull()`

Sollace replied:
Never really used that before...

Sollace replied:
Oh. This is actually part of Java? I always through Objects/Lists/Maps/etc were strictly part of Guava.

mattmess1221 replied:
Guava's version is `Preconditions.checkNotNull()`. Apache uses `Validate.nonNull()`.

Sollace replied:
Weird. i would've avoided them simply for the sake of not depending on those libs.

Daomephsta replied:
There's some really useful stuff in the default MC libraries, I wouldn't avoid using them, you're just making your life so much harder. The only benefit is that the code would be portable to another program, which this already isn't. Guava in particular is absolutely packed with useful stuff.

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/KeyCategory.java:
```diff
@@ -0,0 +1,26 @@
+
+package net.fabricmc.fabric.api.client.keybinding;
+
+public interface KeyCategory {
```
This should be a `public final` class with a private constructor, holding explicitly `public static final` constants. Fabric's API does not use interfaces for providing a group of constants.

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +27,87 @@
+					.id(id)
+					.type(type)
+					.code(code)
+					.category(category);
```
This chain is short enough to be declared in a single line

Sollace replied:
I'll defer this one to @killjoy1121. All I know is he tells me they have to be on separate lines otherwise kittens and puppies get hurt.

mattmess1221 replied:
You think I'm the one in charge? You're the one who brought me here.

mattmess1221 commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +27,87 @@
-			this.binding = binding;
+		private InputUtil.Type type = InputUtil.Type.KEYSYM;
+
+		private Identifier id = new Identifier("fabric", "unnamed");
```
I don't suggest giving this a default.

Sollace replied:
So... null then?

ChloeDawn replied:
Yes

mattmess1221 commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/KeyCategory.java:
```diff
@@ -0,0 +1,26 @@
+
+package net.fabricmc.fabric.api.client.keybinding;
+
+public interface KeyCategory {
```
Could these be better as static constants in `FabricKeyBinding`?

Sollace replied:
I didn't want to clutter that class up too much with constants. I _guess_ I could technically more it.

mattmess1221 replied:
Though making it a final class with a private constructor would also work.

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +27,87 @@
+		private Identifier id = new Identifier("fabric", "unnamed");
+
+		private int code;
+
```
This field should be initialized with `GLFW.GLFW_KEY_UNKNOWN`

InsomniaKitten commented:
Technically the `Identifier` is not an ID, this should probably be clarified here. The only thing treating it as an ID is Fabric... It is just a translation key.

Sollace commented:
@InsomniaKitten I get that. it's a bit of a double-edged sword, though. If you intend on taking it as an id, you should treat it like one. If you're not, then it's better to leave it like a string.

I've considered to continue with what I've been told Fabric heavily prefers.

mattmess1221 commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +29,92 @@
-			this.binding = binding;
+		private InputUtil.Type type = InputUtil.Type.KEYSYM;
+
+		private Identifier id = null;
```
`= null` is redundant.

Sollace replied:
I'm keeping the explicit declaration so it's clear it's intended to be null at that point.

asiekierka commented:
Looks good to me.

Sollace commented:
Can we _please_ get #248 merged? I need to rebase this branch to fix conflicts, and I can't do that until this mess is cleaned up.

![image](https://user-images.githubusercontent.com/6429283/59968166-74fc2e00-9535-11e9-8c9d-2273d1774e6c.png)

shedaniel commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -27,23 +31,91 @@
+		    Objects.requireNonNull(id, "Keybindings should be created with an identifier.");
+
+		    if (code == GLFW.GLFW_KEY_UNKNOWN) {
+		        throw new IllegalStateException("Keybingings need a default keycode.");
```
ok wait a second why am i not allowed to make `unbound` keybinds, this is going to break my mods so hard

LemmaEOF replied:
yeah, you should be able to default to unbound.

Sollace replied:
Okay, remove the throw so you can default to unbound.

Sollace commented:
Rebased to resolve conflicts.

Sollace commented:
> 14 days ago

Is everything still all good?

Hypatech commented:
Any update?

Sollace commented:
>  26 days ago

Guys. Please.

liach commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -41,18 +61,66 @@ protected FabricKeyBinding(Identifier id, InputUtil.Type type, int code, String
+			return new FabricKeyBinding(id, type, code, category);
+		}
+
+		public static Builder create() {
```
can you move this to FabricKeybinding.builder instead of FabricKeybindingBuilder.create?

Sollace replied:
Isn't it the norm to do `FabricKeybinding.Builder.create()....build()`?

Unfortunately, I didn't have the luxury to name the original builder class...

Pyrofab replied:
I most often see `FabricKeyBinding.builder()...build()`. That's the pattern adopted by guava for example, as well as the dimensions API.

Sollace replied:
Alrighty. I'll switch to that.

liach commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/KeyCategory.java:
```diff
@@ -0,0 +1,29 @@
+
+package net.fabricmc.fabric.api.client.keybinding;
+
+public final class KeyCategory {
```
KeyCategories

liach commented:
also @sollace change the pr base branch to 1.14 or 1.15

Pyrofab commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -41,18 +61,66 @@ protected FabricKeyBinding(Identifier id, InputUtil.Type type, int code, String
+		/**
+		 * @deprecated Prefer using the builder itself.
+		 * <pre><code>
+		 * FabricKeyBinding.Builder builder = new FabricKeyBinding.Builder();
```
This javadoc is outdated considering the private constructor.

Pyrofab commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -41,18 +61,66 @@ protected FabricKeyBinding(Identifier id, InputUtil.Type type, int code, String
+		 *				.build();
+		 * KeyBindingRegistry.register(left);
+		 * KeyBindingRegistry.register(right);
+		 * </code></pre>
```
This whole example should probably be on the class javadoc instead of the deprecated method.

Pyrofab commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -41,18 +61,66 @@ protected FabricKeyBinding(Identifier id, InputUtil.Type type, int code, String
+		@Deprecated
 		public static Builder create(Identifier id, InputUtil.Type type, int code, String category) {
-			return new Builder(new FabricKeyBinding(id, type, code, category));
+			return new Builder().id(id).type(type).code(code).category(category);
```
This should call the builder factory method, to avoid confusion.

Pyrofab commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -28,8 +32,24 @@
+		this.id = id;
+	}
+
+	public static String formatKeyName(final Identifier id) {
```
Why is this the only place where the parameter is final ?

Sollace replied:
Presumably because `id`. I'll remove the final.

Sollace commented:
@liach Should be all set! Message me when you want to merge and I'll squash it.

Sollace commented:
@liach Oh, well okay. :/

My statement applies to whoever does end up merging it. Didn't mean to direct iy as you, @liach .

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/KeyCategories.java:
```diff
@@ -0,0 +1,29 @@
+	public static final String GAMEPLAY =	"key.categories.gameplay";
+	public static final String MULTIPLAYER = "key.categories.multiplayer";
+	public static final String UI =		  "key.categories.ui";
+	public static final String MISC =		"key.categories.misc";
```
This is a mix of tabs and spaces

Sollace replied:
Oh what. Thanks, I'll fix that now.

Hypatech commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -26,10 +30,40 @@
+ * <pre><code>
+ * FabricKeyBinding.Builder builder = new FabricKeyBinding.builder();
+ * FabricKeyBinding left = builder
+ *			.id(new identifier("example", "left"))
```
A little nitpicky, but this should be capitalized :)
```suggestion
 *			.id(new Identifier("example", "left"))
```

Hypatech commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -26,10 +30,40 @@
+ *			.code(Keys.Left)
+ *			.build();
+ * FabricKeyBinding right = builder
+ *			.id(new identifier("example", "right"))
```
Same here
```suggestion
 *			.id(new Identifier("example", "right"))
```

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -40,19 +74,54 @@ protected FabricKeyBinding(Identifier id, InputUtil.Type type, int code, String
+
+		private String category = KeyCategories.MISC;
+
+		private Builder() { }
```
```suggestion
		private Builder() {
		}
```

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/KeyCategories.java:
```diff
@@ -0,0 +1,29 @@
+	public static final String UI =          "key.categories.ui";
+	public static final String MISC =        "key.categories.misc";
+
+	private KeyCategories() { }
```
```suggestion
	private KeyCategories() {
	}
```

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -26,10 +30,40 @@
  * {@link KeyBindingRegistry#register(FabricKeyBinding)}!
+ * </p>
+ * <pre><code>
+ * FabricKeyBinding.Builder builder = new FabricKeyBinding.builder();
```
Is it a good idea to encourage reuse of a builder type?

Sollace replied:
Only did that for the example, but there is no reason why you shouldn't. The state of the builder is in no way linked to the keys you create with it, it simply holds the parameters, and there are benefits to being able to create a builder, use it to create a key, change one/two of its parameters, and create another key.

If you're _not_ planning to reuse the builder, it had might as well be an over-glorified constructor.

ChloeDawn commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -26,10 +30,40 @@
+
+	/**
+	 * Original identifier used to register this key.
+	 * May be different from the {@link getId()}!
```
Documentation should be formal and descriptive. This should also explain *why* they can differ.
```suggestion
	 * May be different from the {@link getId()}
```

natanfudge replied:
Really, having the exclamation mark is more descriptive because it bundles the metadata of the feeling of the sentence.

Sollace replied:
I don't feel like there's really a need to explain why they differ. If you really want to know: It's because Fabric requires you to pass in an identifier, but doesn't actually use it. Instead it converts and mutilates it into a string that gets passed down as the _actual_ identifier.

For a key created with the Identifier `example:mod`, calling `getId()` and parsing that to an Identifier will return back `minecraft:key.example.mod` (hence why I hate that design decision).

ChloeDawn replied:
The root problem is Yarn calling it an ID when it's a translation key lol

Sollace commented:
Squashed and updated to 1.15. Also added some new features/configurability to the builder.

I'd like to make this into a fabric-keybindings-v1 package, though, so we can remove the backwards compat.

Player3324 commented:
The PR has some checkstyle violations and adding new classes in the wrong api package is probably not what we want to do. I think we want to migrate the module to v1, fixing the api package, like we did for https://github.com/FabricMC/fabric/pull/359

Pyrofab commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -41,19 +72,148 @@ protected FabricKeyBinding(Identifier id, InputUtil.Type type, int code, String
+		 */
 		public FabricKeyBinding build() {
+			Objects.requireNonNull(id, "Keybindings should be created with an identifier.");
+			Preconditions.checkState(unassigned || code != UNASSIGNED, "Keybingings need a default keycode.");
```
```suggestion
			Preconditions.checkState(unassigned || code != UNASSIGNED, "Keybindings need a default keycode.");
```

frqnny commented:
If this is dead, why not close it?

liach commented:
Closing in favor of #615
